framework/src/csg/CSGSurface.C

```diff
@@ -0,0 +1,51 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "CSGSurface.h"
+
+namespace CSG
+{
+
+CSGSurface::CSGSurface(const std::string & name, const std::string & surf_type)
+  : _name(name), _surface_type(surf_type)
+{
+}
+
+CSGSurface::Halfspace
+CSGSurface::getHalfspaceFromPoint(const Point & p) const
+{
+  auto eval = evaluateSurfaceEquationAtPoint(p);
+  if (MooseUtils::absoluteFuzzyGreaterThan(eval, 0))
+    return Halfspace::POSITIVE;
+  else if (MooseUtils::absoluteFuzzyLessThan(eval, 0))
+    return Halfspace::NEGATIVE;
+  else
+    mooseError("Point ",
+               p,
+               " used to determine halfspace evaluation lies on the surface ",
+               _name,
+               ", leading to an ambiguously defined halfspace.");
+}
+
+bool
+CSGSurface::operator==(const CSGSurface & other) const
+{
+  const auto name_eq = this->getName() == other.getName();
+  const auto surface_type_eq = this->getSurfaceType() == other.getSurfaceType();
+  const auto coeffs_eq = this->getCoeffs() == other.getCoeffs();
+  return (name_eq && surface_type_eq && coeffs_eq);
```

Comment:
It's actually better to do it like this so that you can get early returns (although the compiler might just optimize it away):

```
return (this->getName() == other.getName()) && this->getSurfaceType() == other.getSurfaceType() ...
```